src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/TraversingBlockBalancedTreeReader.java

```diff
@@ -0,0 +1,314 @@
+        private final int[] rightNodePositions;
+        // true if the per-dim delta we read for the node at this level is a negative offset vs. the last split on this dim; this is a packed
+        // 2D array, i.e. to access array[level][dim] you read from negativeDeltas[level*numDims+dim].  this will be true if the last time we
+        // split on this dimension, we next pushed to the left subtree:
+        private final boolean[] negativeDeltas;
+        // holds the packed per-level split values; the run method uses this to save the cell min/max as it recurses:
+        private final byte[][] splitValuesStack;
+        // scratch value to return from getPackedValue:
+        private final BytesRef scratch;
+
+        PackedIndexTree()
+        {
+            int treeDepth = getTreeDepth();
+            splitPackedValueStack = new byte[treeDepth + 1][];
+            nodeID = 1;
+            level = 1;
+            splitPackedValueStack[level] = new byte[packedBytesLength];
+            leafBlockFPStack = new long[treeDepth + 1];
+            leftNodePositions = new int[treeDepth + 1];
+            rightNodePositions = new int[treeDepth + 1];
+            splitValuesStack = new byte[treeDepth + 1][];
+            negativeDeltas = new boolean[treeDepth + 1];
+
+            in = new ByteArrayDataInput(packedIndex);
+            splitValuesStack[0] = new byte[packedBytesLength];
+            readNodeData(false);
+            scratch = new BytesRef();
+            scratch.length = bytesPerValue;
+        }
+
+        public void pushLeft()
+        {
+            int nodePosition = leftNodePositions[level];
+            nodeID *= 2;
+            level++;
+            if (splitPackedValueStack[level] == null)
+                splitPackedValueStack[level] = new byte[packedBytesLength];
+            System.arraycopy(negativeDeltas, level - 1, negativeDeltas, level, 1);
+            assert !leafNode;
+            negativeDeltas[level] = true;
+            in.setPosition(nodePosition);
+            readNodeData(true);
+        }
+
+        public void pushRight()
+        {
+            int nodePosition = rightNodePositions[level];
+            nodeID = nodeID * 2 + 1;
+            level++;
+            if (splitPackedValueStack[level] == null)
+                splitPackedValueStack[level] = new byte[packedBytesLength];
+            System.arraycopy(negativeDeltas, level - 1, negativeDeltas, level, 1);
+            assert !leafNode;
+            negativeDeltas[level] = false;
+            in.setPosition(nodePosition);
+            readNodeData(false);
+        }
+
+        public void pop()
+        {
+            nodeID /= 2;
+            level--;
+            leafNode = false;
+        }
+
+        public boolean isLeafNode()
+        {
+            return nodeID >= numLeaves;
+        }
+
+        public boolean nodeExists()
+        {
+            return nodeID - numLeaves < numLeaves;
+        }
```

Comment:
nit: I'm 50/50 on whether it makes sense for `isLeafNode()` and `nodeExists()` to remind the reader via JavaDoc that the calculations work because the tree is balanced :)